### PR TITLE
Feature/multi language support grading buttons

### PR DESCRIPTION
**Pull request by Cora47**
### Checklist
- [ ] I tested *all* changes and their related features locally or with a test instance of Artemis.
- [x] I documented the Java code using JavaDoc style.
- [x] I documented the changes in the Documentation (/docs).

### Motivation and Context

To enable the graders to better work with Artemis, the buttons depicting and the corresponding descriptions can now be viewed in multiple languages. 

This corresponds to issue https://github.com/kit-sdq/programming-lecture-eclipse-artemis/issues/259.

### Description

The json file is now expecting not a string for name and message but a dictionary containing all names and messages as following: 

```
{
    "shortName": "jdEmpty",
    "button": {
        "de": "JavaDoc Leer",
        "en": "JavaDoc Empty"
    },
    "message": {
        "de": "JavaDoc ist leer oder nicht vorhanden",
        "en": "JavaDoc is empty or does not exist"
    },
    "penaltyRule": {
        "shortName": "thresholdPenalty",
        "threshold": 1,
        "penalty": 5
    },
    "appliesTo": "style"
},
```

If the needed default message is not in the dictionary the JSON file provides, an IOException is thrown, which will be catched in the Constructor of the AssesmentController (package edu.kit.kastel.eclipse.common.core;). As the default name is still needed for the program to function correctly, I included this in the conditions for the error.

The keys in the JSON Dictionary must correspond to [ISO 639 alpha-2 or alpha-3 language code](https://de.wikipedia.org/wiki/Liste_der_ISO-639-1-Codes), therefore it is easily expandable in regard to adding new languages to buttons or messages.

### Steps for Testing

As the Test-Server is down at the moment, I am unsure as to how to test all changes. I did however add a few JSON files for testing in docs/examples. These have to be referred to by name in the Eclipse Test Instance Setting - Artemis - Configuration File. Loading of the configuration is triggered when loading a specific project to begin marking.

-   config_v6.json is just the edited format with only the German button names and messages in the new format
-   config_v6_buttons_dicts_short.json contains fewer buttons, but with the possibility of showing the English variants. I tested this previously and it worked.
-   config_v6_buttons_error_missing_default.json is for triggering the aforementioned error. It only misses a default message and is based on the second file

### CODE REVIEW

**Comment by dfuchss** [comment]
@cora47 please resolve the merging issues before review :)

**Comment by Cora47** [comment]
@dfuchss merging should be done. Is there still something off?

File: bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/MistakeType.java [hunk 1]
```diff
@@ -28,17 +36,43 @@ public class MistakeType implements IMistakeType {
...
+		// locale -> getCountry
+
+		names = new HashMap<>();
+		this.messages = new HashMap<>();
+
+		if (messages != null) {
+			Set<String> messageKeys = messages.keySet();
+			for (String key : messageKeys) {
+				this.messages.put(new Locale(key), messages.get(key));
+			}
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/MistakeType.java:59]: Iterate over the entry set

File: bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java [hunk 2]
```diff
@@ -18,6 +18,8 @@ public class RatingGroup implements IRatingGroup {
 	@JsonProperty
 	private String displayName;
 	@JsonProperty
+	private String engDisplayName;
+	@JsonProperty
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java:22]: Remove the changes to the rating group for now. Important are the mistakes. In addition, here the languages are hard coded

**Review by dfuchss — CHANGES_REQUESTED** [review]
I've added some remarks to the code.

Did you tested that in English mode, the data sent to Artemis still contains the German texts?

**Comment by dfuchss** [comment]
@cora47 merging was successful. I've added some remarks regarding the code here.

File: bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java [hunk 3]
```diff
@@ -37,6 +39,22 @@ public String getDisplayName() {
 		return this.displayName;
 	}
 
+	@Override
+	/**
+	 * @since 2.7
+	 */
+	public String getLanguageSensitiveDisplayName(String language) {
+		// TODO Auto-generated method stub
```
Comment (tsaglam) [inline bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java:47]: leftover generated todo

File: bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java [hunk 4]
```diff
@@ -37,6 +39,22 @@ public String getDisplayName() {
...
+		switch (language) {
+		case "English":
+			return engDisplayName;
+		case "Deutsch":
```
Comment (tsaglam) [inline bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java:51]: Language strings should either be constants or enum.

Comment (tsaglam) [reply bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java:51]: Or can you derive these values from the locale?

**Review by tsaglam — CHANGES_REQUESTED** [review]
Just a few minor things in addition to the feedback of @dfuchss.

**Comment by tsaglam** [comment]
@Cora47 can you please merge the current main branch again and then click resolve for all review comments that you already addressed, so we can see what is left to to/discuss.

File: bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/MistakeType.java (hunk 1 above)
Comment (Cora47) [reply bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/MistakeType.java]: Done

File: bundles/edu.kit.kastel.sdq.eclipse.common.api/src/edu/kit/kastel/eclipse/common/api/model/IRatingGroup.java [hunk 5]
```diff
@@ -15,6 +15,15 @@
 public interface IRatingGroup {
 	String getDisplayName();
 
+	/**
+	 * Not suitable for official annotations
+	 * 
+	 * @default is the official German name
```
[the remark this answers is not in the document]
Comment (Cora47) [reply bundles/edu.kit.kastel.sdq.eclipse.common.api/src/edu/kit/kastel/eclipse/common/api/model/IRatingGroup.java]: Was something I played around with, but is ultimately unused, so I removed this

File: bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/MistakeType.java [hunk 6]
```diff
@@ -28,17 +36,43 @@ public class MistakeType implements IMistakeType {
...
+		}
+
+		if (buttons != null) {
+			Set<String> nameKeys = buttons.keySet();
+			for (String key : nameKeys) {
+				names.put(new Locale(key), buttons.get(key));
+			}
```
[the remark this answers is not in the document]
Comment (Cora47) [reply bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/MistakeType.java]: done

File: bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java [hunk 7]
```diff
@@ -37,6 +39,22 @@ public String getDisplayName() {
...
+			return displayName;
+		default:
+			return getDisplayName();
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Cora47) [reply bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java]: yeah, code I played around with, as it is unused I removed it

File: bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java (hunk 2 above)
Comment (Cora47) [reply bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java]: done

File: bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/GermanLanguage.java [hunk 8]
```diff
@@ -1,4 +1,4 @@
-/* Licensed under EPL-2.0 2022. */
+/* Licensed under EPL-2.0 2022-2023. */
 package edu.kit.kastel.eclipse.common.view.languages;
 
 /**
```
[the remark this answers is not in the document]
Comment (Cora47) [reply bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/GermanLanguage.java:4]: Done the override

File: bundles/edu.kit.kastel.sdq.eclipse.common.api/src/edu/kit/kastel/eclipse/common/api/model/IRatingGroup.java [hunk 9]
```diff
@@ -15,6 +15,15 @@
...
+	 * @since 2.7
```
[the remark this answers is not in the document]
Comment (Cora47) [reply bundles/edu.kit.kastel.sdq.eclipse.common.api/src/edu/kit/kastel/eclipse/common/api/model/IRatingGroup.java]: done

File: bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java (hunk 3 above)
Comment (Cora47) [reply bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java]: is removed

File: bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java (hunk 4 above)
Comment (Cora47) [reply bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/RatingGroup.java]: as ratingGroup changes are reversed, comment is now irrelevant, but was done like this in MistakeType.java

File: bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/MistakeType.java [hunk 10]
```diff
@@ -28,17 +36,45 @@ public class MistakeType implements IMistakeType {
 	 * having to add
 	 * <li>the rating group
 	 * <li>this object to the rating Group
+	 * 
+	 * @throws IOException if the JSON File is not build correctly
+	 * @since 2.9
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/MistakeType.java:41]:
```suggestion
```

File: bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/MistakeType.java [hunk 11]
```diff
@@ -13,9 +18,12 @@
 import edu.kit.kastel.eclipse.common.core.model.rule.PenaltyRule;
 
 public class MistakeType implements IMistakeType {
+	private static final Locale DEFAULT_LOCALE = Locale.GERMAN;
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/MistakeType.java:21]: Add this to the Class ArtemisConstants (in the API Module (as far as I remember))

File: bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/MistakeType.java [hunk 12]
```diff
@@ -28,17 +36,45 @@ public class MistakeType implements IMistakeType {
...
+			@JsonProperty("appliesTo") String appliesTo) throws IOException {
+
 		this.shortName = shortName;
-		this.name = buttonName;
-		this.message = message;
-		this.penaltyRule = penaltyRule;
 
+		// locale -> getCountry
+
+		names = new HashMap<>();
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.common.core/src/edu/kit/kastel/eclipse/common/core/model/MistakeType.java:52]:
```suggestion
		this.names = new HashMap<>();
```

File: bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/GermanLanguage.java [hunk 13]
```diff
@@ -390,4 +392,10 @@ public String assessment() {
 	public String languageDisplayName() {
 		return "Deutsch";
 	}
+
+	@Override
+	public Locale getLocale() {
+		// TODO Auto-generated method stub
+		return Locale.GERMAN;
+	}
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.common.view/src/edu/kit/kastel/eclipse/common/view/languages/GermanLanguage.java:400]: Cleanup TODOs

File: docs/grading/grading.dev.rst [hunk 14]
```diff
@@ -152,8 +152,14 @@ Creating a new PenaltyRule
...
                 "shortName": "idk",
-                "button": "MyMistakeType",
-                "message": "You made a grave mistake.",
+                "button": {
+		    "de": "MeinFehlerTyp",
+		    "en": "MyMistakeType"
+	    	},
+                "message": {
+		    "de": "Du machst einen schweren Fehler",
+		    "en": "You made a grave mistake."
```
Comment (dfuchss) [inline docs/grading/grading.dev.rst:161]:
```suggestion
		    "en": "You made a big mistake."
```

**Review by dfuchss — CHANGES_REQUESTED** [review]

**Pull request closed without merging**